File: src/middle/interpret.jl
function interpret(int, ir::IR, args...)
  env = Dict{Variable,Any}()
  resolve(x) = x
  resolve(v::Variable) = env[v]
  resolve(v::Binding) = resolve_global(int.defs, v)
  foreach(((v, x),) -> env[v] = x, zip(arguments(ir), args))
  bl = 1
  while true
    for (v, st) in block(ir, bl)
      args = resolve.(st.expr.args)
      any(==(⊥), args) && return
      if isexpr(st, :call)
        if (w = args[1]) isa WIntrinsic
          haskey(wasmPartials, w.op) || return
          args = args[2:end]
          env[v] = all(isvalue, args) ? wasmPartials[w.op](args...) : rvtype(w.ret)
        else
          result = int[(args...,)]
          isnothing(result) && return
          env[v] = result
        end
      elseif isexpr(st, :pack)
        env[v] = pack(args...)
      elseif isexpr(st, :branch)
        if isreturn(st.expr)
          return args[3]
        elseif isunreachable(st.expr)
          return
        else
          target, cond, args... = args
          if !isnothing(cond)
            cond isa Int32 || return
            Bool(cond) || continue
          end
          bl = target
          foreach(((v, x),) -> env[v] = x, zip(arguments(block(ir, bl)), args))
          break
        end
      elseif isexpr(st, :(=))
        return
      else
        error("Unknown expr type $(st.expr.head)")
      end
    end
  end
end

function interpret(int, meth::RMethod, args...)
  meth == invoke_method && return
  meth.func isa IR ? interpret(int, meth.func, args...) : meth.func(args...)
end

function interpret(int, func::Tag, args)
  for meth in reverse(int.defs.methods[func])
    m = partial_match(int, meth.sig.pattern, args)
    isnothing(m) && continue
    ismissing(m) && return
    args = [m[a][1] for a in meth.sig.args]
    result = interpret(int, meth, args...)
    isnothing(result) && return
    return isempty(meth.sig.swap) ? rlist(result) : result
  end
end

struct Interpreter
  defs::Definitions
  results::Cache{Any,Any}
end

function Interpreter(defs::Definitions)
  ch = Cache{Any,Any}() do self, sig
    self = Interpreter(defs, self)
    interpret(self, sig...)
  end
  return Interpreter(defs, ch)
end

Base.getindex(int::Interpreter, sig) = int.results[sig]

Caches.subcaches(int::Interpreter) = (int.results,)

File: src/middle/interpret.ts
import * as types from '../frontend/types.js'
import { Tag, Type } from '../frontend/types.js'
import { Method, Definitions, IRValue, MIR, Global, SetGlobal, Invoke, Wasm, Value } from '../frontend/modules.js'
import { IR, unreachable, Branch, asType } from '../utils/ir.js'
import { CycleCache } from '../utils/cache.js'
import { partial_match } from './patterns.js'
import { wasmPartials } from '../backend/wasm.js'
import { invoke_method } from './primitives.js'
import isEqual from 'lodash/isEqual.js'
import { some } from '../utils/map.js'
import { Instruction } from '../wasm/wasm.js'

export { Interpreter, interpret, interpreter }

type Func = Tag | Method

class Interpreter {
  readonly results: { get: (k: [Func, ...Type[]]) => Type | undefined }

  constructor(readonly defs: Definitions, results?: { get: (k: [Func, ...Type[]]) => Type | undefined }) {
    if (results) {
      this.results = results
    } else {
      this.results = new CycleCache<[Func, ...Type[]], Type | undefined>(_ => undefined, (self, sig) => {
        const int = new Interpreter(defs, self)
        return interpret(int, ...sig)
      })
    }
  }

  get(func: Func, args: Type[]): Type | undefined { return this.results.get([func, ...args]) }
}

function interpret(int: Interpreter, x: Func, ...args: Type[]): Type | undefined {
  if (x instanceof IR) return interpretIR(int, x, ...args)
  if (x instanceof Method) return interpretMethod(int, x, ...args)
  if (args.length !== 1) throw new Error(`Expected 1 argument, got ${args.length}`)
  return interpretFunc(int, x, args[0])
}

function interpretIR(int: Interpreter, ir: MIR, ...args: Type[]): Type | undefined {
  const env = new Map<number, Type>()
  const resolve = (x: number | IRValue): Type => {
    if (x instanceof Value) throw new Error('unsupported')
    return typeof x === 'number' ? some(env.get(x)) : x
  }
  for (const [v, x] of ir.block(1).args.map((v, i) => [v, args[i]] as const)) env.set(v, x)
  let bl = 1
  while (true) {
    for (const [v, st] of ir.block(bl)) {
      const xs = st.expr.body.map(resolve)
      if (st.expr instanceof Invoke) {
        const result = int.get(st.expr.method, xs)
        if (result === undefined) return
        env.set(v, result)
      } else if (st.expr instanceof Wasm) {
        if (st.expr.isImport()) return
        const instr = st.expr.callee as Instruction
        const op = instr.kind === 'op' ? instr.name : ''
        if (!wasmPartials.has(op)) return
        const args = xs
        const allValues = args.every(types.isValue)
        env.set(v, allValues ? wasmPartials.get(op)!(...args) : asType(st.type))
      } else if (st.expr.head === 'call') {
        const op = xs[0]
        if (!(op instanceof Tag)) throw new Error(`Expected function or method, got ${op}`)
        const result = int.get(op, xs.slice(1))
        if (result === undefined) return
        env.set(v, result)
      } else if (st.expr.head === 'pack') {
        env.set(v, types.pack(...xs))
      } else if (st.expr instanceof Branch) {
        if (st.expr.isreturn()) return xs[0]
        if (st.expr.isunreachable()) return
        if (st.expr.isconditional()) {
          const cond = xs[0]
          if (!types.issubset(cond, types.bool())) throw new Error(`Expected boolean condition`)
          if (!types.isValue(cond)) return
          if (!isEqual(cond, types.bool(true))) continue
        }
        bl = st.expr.target
        let as = st.expr.args.map(x => resolve(x))
        ir.block(st.expr.target).args.forEach((arg, i) => env.set(arg, as[i]))
        break
      } else if (st.expr.head === 'tuple') {
        const T = asType(st.type)
        if (!types.isValue(T)) throw new Error('assert isvalue(st.type)')
        env.set(v, T)
      } else if (st.expr instanceof Global) {
        env.set(v, asType(int.defs.resolve_static(st.expr.binding)))
      } else if (st.expr instanceof SetGlobal) {
        return
      } else {
        throw new Error(`Unknown expr type ${st.expr.head}`)
      }
    }
  }
}

function interpretMethod(int: Interpreter, meth: Method, ...args: Type[]): Type | undefined {
  if (meth === invoke_method) return
  if (meth.func) {
    const result = meth.func(...args)
    return result === unreachable ? undefined : result
  }
  const ir = some(int.defs.ir(meth))
  return interpretIR(int, ir, ...args)
}

function interpretFunc(int: Interpreter, func: Tag, args: Type): Type | undefined {
  const methods = int.defs.methods(func)
  for (const meth of methods.slice().reverse()) {
    const m = partial_match(int, meth.sig.pattern, args)
    if (m === null) continue
    if (m === undefined) return
    const as = meth.sig.args.map(a => m.get(a)![0])
    const result = interpret(int, meth, ...as)
    if (result === undefined) return
    return meth.sig.swap.size === 0 ? types.list(result) : result
  }
}

function interpreter(defs: Definitions): Interpreter { return new Interpreter(defs) }
